crates/red_knot/src/semantic/types/infer.rs

```diff
@@ -68,6 +89,33 @@ where
     }
 }
 
+/// Infer type for a ConstrainedDefinition (intersection of the definition type and the
+/// constraints)
+#[tracing::instrument(level = "trace", skip(db))]
+pub fn infer_constrained_definition_type(
+    db: &dyn SemanticDb,
+    symbol: GlobalSymbolId,
+    constrained_definition: ConstrainedDefinition,
+) -> QueryResult<Type> {
+    let ConstrainedDefinition {
+        definition,
+        constraints,
+    } = constrained_definition;
+    let definition_type = infer_definition_type(db, symbol, definition)?;
+    let constraint_types: Vec<Option<Type>> = constraints
+        .into_iter()
+        .map(|constraint| infer_constraint_type(db, symbol, constraint))
+        .collect::<QueryResult<Vec<_>>>()?;
```

Comment:
Collecting here and for `intersected_types` seems unnecessary. Can we just collect once (e.g. by using a plain old loop ;))